bench/src/main/scala/play/api/libs/ws/ahc/AddQueryParamsBench.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.api.libs.ws.ahc
+
+import java.util.concurrent.TimeUnit
+
+import akka.stream.Materializer
+import org.openjdk.jmh.annotations._
+import org.openjdk.jmh.infra.Blackhole
+import play.api.libs.ws.StandaloneWSRequest
+
+/**
+ * ==Quick Run from sbt==
+ *
+ * > bench/jmh:run .*AddQueryParamsBench
+ *
+ * ==Using Oracle Flight Recorder==
+ *
+ * To record a Flight Recorder file from a JMH run, run it using the jmh.extras.JFR profiler:
+ * > bench/jmh:run -prof jmh.extras.JFR .*AddQueryParamsBench
+ *
+ * Compare your results before/after on your machine. Don't trust the ones in scaladoc.
+ *
+ * Sample benchmark results:
+ * {{{
+ * > bench/jmh:run .*AddQueryParamsBench
+ * [info] Benchmark                      (size)  Mode  Cnt    Score   Error  Units
+ * [info] AddQueryParamsBench.addParams       1  avgt   10   47.978 ± 0.388  ns/op
+ * [info] AddQueryParamsBench.addParams      10  avgt   10   76.495 ± 0.663  ns/op
+ * [info] AddQueryParamsBench.addParams     100  avgt   10  118.205 ± 1.072  ns/op
+ * [info] AddQueryParamsBench.addParams    1000  avgt   10  141.506 ± 1.681  ns/op
+ * [info] AddQueryParamsBench.addParams   10000  avgt   10  151.452 ± 2.560  ns/op
+ * }}}
+ *
+ * @see https://github.com/ktoso/sbt-jmh
+ */
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@BenchmarkMode(Array(Mode.AverageTime))
+@Fork(jvmArgsAppend = Array("-Xmx350m", "-XX:+HeapDumpOnOutOfMemoryError"), value = 1)
+@State(Scope.Benchmark)
+class AddQueryParamsBench {
```

Comment:
Can we also have a test for the header changes? It would be good if we have tests for both `with*` and `add*` methods.

Reply:
Of course.

Reply:
I've added benchmarks for `addHttpHeaders` but stopped short of adding them for `with*` since it doesn't do anything interesting.

Since these are run manually, they're not going to catch perf regressions. For that, you'd need a bigger investment to to hook them into the build and have consistent (ideally dedicated) hardware. Personally, I like micro-benchmarks are useful for validating before/after changes to interesting parts.

I'll add defer and add them to `with*` if you really want to, but I'm not convinced there's much to be gained from it. An E2E benchmark for typical request construction might be interesting to further micro-optimize the library, but I'd suggest it's beyond the scope of this PR.

Reply:
Yep, we are not running them in dedicated hardware yet. But we have this for playframework, running nightly. We can later add a nightly script to run the benchmarks here, but I agree with you that we can add them to `with*` methods then.